src/NuGet.Core/NuGet.Common/PathUtil/DirectoryUtility.cs

```diff
@@ -42,7 +41,15 @@ public static void CreateSharedDirectory(string path)
                     var currentPath = sepPos == -1 ? path : path.Substring(0, sepPos);
                     if (!Directory.Exists(currentPath))
                     {
-                        CreateSingleSharedDirectory(currentPath);
+                        // There are potential race conditions when multiple threads are trying to create the same shared path.
+                        // This simple lock ensures that we are consistent.
+                        lock (LockObject)
+                        {
+                            if (!Directory.Exists(currentPath))
+                            {
+                                CreateSingleSharedDirectory(currentPath);
```

Comment:
I'd not bother with all of the ACL-then-move if you can get away with it. Just create and ACL if you're already locked.

Reply:
I misunderstood your point in the e-mail. 
Fixed.